ext/datadog_profiling_native_extension/heap_recorder.c

```diff
@@ -951,112 +875,30 @@ st_index_t heap_stack_hash(heap_stack *stack, st_index_t seed) {
-  if (key->type == HEAP_STACK) {
-    char *filename = key->heap_stack->frames[entry_i].filename;
-    (*filename_ptr) = filename;
-    return strlen(filename);
-  } else {
-    ddog_CharSlice filename = key->location_slice->ptr[entry_i].function.filename;
-    (*filename_ptr) = filename.ptr;
-    return filename.len;
-  }
-}
-
-int heap_record_key_cmp_st(st_data_t key1, st_data_t key2) {
-  heap_record_key *key_record1 = (heap_record_key*) key1;
-  heap_record_key *key_record2 = (heap_record_key*) key2;
+int heap_stack_cmp_st(st_data_t key1, st_data_t key2) {
+  heap_stack *stack1 = (heap_stack*) key1;
+  heap_stack *stack2 = (heap_stack*) key2;
 
   // Fast path, check if lengths differ
-  size_t key_record1_len = heap_record_key_len(key_record1);
-  size_t key_record2_len = heap_record_key_len(key_record2);
-
-  if (key_record1_len != key_record2_len) {
-    return ((int) key_record1_len) - ((int) key_record2_len);
+  if (stack1->frames_len != stack2->frames_len) {
+    return ((int) stack1->frames_len) - ((int) stack2->frames_len);
   }
 
   // If we got this far, we have same lengths so need to check item-by-item
-  for (size_t i = 0; i < key_record1_len; i++) {
-    // Lines are faster to compare, lets do that first
-    size_t line1 = heap_record_key_entry_line(key_record1, i);
-    size_t line2 = heap_record_key_entry_line(key_record2, i);
-    if (line1 != line2) {
-      return ((int) line1) - ((int)line2);
-    }
+  for (size_t i = 0; i < stack1->frames_len; i++) {
+    heap_frame* frame1 = &stack1->frames[i];
+    heap_frame* frame2 = &stack2->frames[i];
 
-    // Then come names, they are usually smaller than filenames
-    const char *name1, *name2;
-    size_t name1_len = heap_record_key_entry_name(key_record1, i, &name1);
-    size_t name2_len = heap_record_key_entry_name(key_record2, i, &name2);
-    if (name1_len != name2_len) {
-      return ((int) name1_len) - ((int) name2_len);
-    }
-    int name_cmp_result = strncmp(name1, name2, name1_len);
-    if (name_cmp_result != 0) {
-      return name_cmp_result;
+    if (frame1->name != frame2->name) {
+      return ((int) frame1->name) - ((int) frame2->name);
     }
 
-    // Then come filenames
-    const char *filename1, *filename2;
-    int64_t filename1_len = heap_record_key_entry_filename(key_record1, i, &filename1);
-    int64_t filename2_len = heap_record_key_entry_filename(key_record2, i, &filename2);
-    if (filename1_len != filename2_len) {
-      return ((int) filename1_len) - ((int) filename2_len);
+    if (frame1->filename != frame2->filename) {
+      return ((int) frame1->filename) - ((int) frame2->filename);
     }
-    int filename_cmp_result = strncmp(filename1, filename2, filename1_len);
-    if (filename_cmp_result != 0) {
-      return filename_cmp_result;
+
+    if (frame1->line != frame2->line) {
+      return ((int) frame1->line) - ((int)frame2->line);
     }
   }
```

Comment:
As an optimization here, now that this is all just a bunch of numbers, maybe we can just `memcmp` the two stacks? (I think this would be a stable comparison similar to the  current implementation?)